presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -561,17 +562,23 @@ protected Scope visitCall(Call node, Optional<Scope> scope)
             return createAndAssignScope(node, scope);
         }
 
-        private void validateColumns(Statement node, RelationType descriptor)
+        private void validateColumns(Statement node, RelationType descriptor, List<String> columnAliases)
         {
             // verify that all column names are specified and unique
             // TODO: collect errors and return them all at once
+            boolean hasColumnAliases = !columnAliases.isEmpty();
+            if (hasColumnAliases && columnAliases.size() != descriptor.getVisibleFieldCount()) {
+                throw new SemanticException(MISMATCHED_COLUMN_ALIASES, node, "Column alias list has %s entries but subquery has %s columns", columnAliases.size(), descriptor.getVisibleFields().size());
+            }
+
             Set<String> names = new HashSet<>();
             for (Field field : descriptor.getVisibleFields()) {
                 Optional<String> fieldName = field.getName();
-                if (!fieldName.isPresent()) {
+                if (!fieldName.isPresent() && !hasColumnAliases) {
                     throw new SemanticException(COLUMN_NAME_NOT_SPECIFIED, node, "Column name not specified at position %s", descriptor.indexOf(field) + 1);
                 }
-                if (!names.add(fieldName.get())) {
+                String columnName = hasColumnAliases ? columnAliases.get(descriptor.indexOf(field)) : fieldName.get();
```

Comment:
> `descriptor.indexOf(field)`

You're iterating over `descriptor.getVisibleFields()`, but you ask for field index. It seems not guaranteed you will get contiguously 0,1,2,3...